src/main/java/saigonwithlove/ivy/intellij/shared/LibrarySyncInvoker.java

```diff
@@ -0,0 +1,45 @@
+package saigonwithlove.ivy.intellij.shared;
+
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.application.ModalityState;
+import com.intellij.openapi.components.ServiceManager;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.ui.MessageDialogBuilder;
+import com.intellij.openapi.ui.Messages;
+import lombok.NoArgsConstructor;
+import org.apache.commons.lang3.StringUtils;
+import org.jetbrains.annotations.NotNull;
+import saigonwithlove.ivy.intellij.engine.IvyEngineService;
+import saigonwithlove.ivy.intellij.settings.PreferenceService;
+
+@NoArgsConstructor
+public class LibrarySyncInvoker {
```

Comment:
I'm sorry for writing many comments. I just want to make the code as consistent as possible, please forgive me about this.

When writing GUI application like this one, we will try to separate GUI code with back-end code. In this case, the GUI code will inside IvySettingView and InitializationActivity, the back-end code will go to IvyEngineService. We will try not to introduce new term. The LibrarySyncInvoker may not needed because its code can split into the above classes.

I can imagine the code will look like this in IvySettingView and InitializationActivity
```
PreferenceService.State preferences = preferenceService.getState();
IvyEngineService ivyEngineService = ServiceManager.getService(project, IvyEngineService.class);
if (ivyEngineService.isOsgiFolderExist(preferences.getIvyEngineDirectory()) {
  ApplicationManager.getApplication().runWriteAction(() -> ServiceManager.getService(project, IvyEngineService.class).addLibraries(preferences.getIvyEngineDirectory()));
} else {
  int result = MessageDialogBuilder.yesNo(..).yesText("Start Engine").noText("Manually start").project(project).show();
  if (result == Messages.YES) {
    ivyEngineService.startIvyEngine(preferences.getIvyEngineDirectory(), project);
  }
}
```

This code is still duplicated but more readable, we don't need to go to another layer to see what the code does.

Notice the messages, those "text" messages will be put in `resources.messages.IvyBundle.properties`, we can do i18n easier.

Reply:
It's fine. I want to learn more, so please don't hesitate to comment. For this point I get your idea but I have one concern. What if user select Setting > choose another directory. Should we also show the warning dialog? If we do then we do same thing with the behaviour at InitializationActivity. Then how can we centralize in this case?

Reply:
Yes, we have to duplicate the block of code above, it should be the same in `IvySettingView` and `InitializationActivity`. The `IvySettingView` handles setting change you mentioned, the `InitializationActivity` will update libraries automatically at startup. I can't find the way to centralize this block that satisfy me, we can accept the code duplication in this case.